### PR TITLE
feat: Share util functions

### PR DESCRIPTION
**Pull request by laurentsimon**
This PR does 3 things:
1. Share util functions between builders
2. Ensure the files we create do not already exist (`O_EXCL`)
3. Add sanity checks when downloading an artifact for the non-existence of the file

I started looking into 

https://github.com/slsa-framework/slsa-github-generator/blob/0f97cca414ee57448ec44667e92aa65c46a23eb6/internal/builders/go/main.go#L87 

and https://github.com/slsa-framework/slsa-github-generator/blob/main/internal/builders/generic/attest.go#L165 

Then I realized that a malicious developer can use this function to sign any file present in the generator directory. The reason is that if the file already exists, this function does not return an error and continue silently. Fortunately we don't checkout the developer repository (by design), because it might have allowed the developer to set the path to a file they control.

I think an attacker could only have selected files with `.intoto.jsonl` extension, so just a few testcase files. I have not tested filenames like `some-file\0`(to remove the extension in https://github.com/slsa-framework/slsa-github-generator/blob/0f97cca414ee57448ec44667e92aa65c46a23eb6/internal/builders/go/main.go#L86), but I doubt it would work...

So I created the `CreateNewFileUnderCurrentDirectory()` function which uses `O_EXCL` and use it across the 2 existing builders.

Besides these changes, the rest is mostly re-factoring.

### CODE REVIEW

**Comment by ianlewis** [comment]
Thanks, I'll give this a review tomorrow

File: internal/builders/go/main_test.go
Comment (asraa) [inline internal/builders/go/main_test.go:270]: nit: leftover debug?

File: internal/utils/marshall.go
Comment (asraa) [inline internal/utils/marshall.go:42]: super small nit: Marshall to Marshal everywhere?

File: internal/utils/marshall.go
Comment (asraa) [inline internal/utils/marshall.go:58]: Did you want to return `nil, err` here?

**Review by asraa — COMMENTED** [review]
Thanks! Really minor nits

File: internal/builders/go/main_test.go
Comment (laurentsimon) [reply internal/builders/go/main_test.go:270]: good catch. Thanks!

File: internal/utils/marshall.go
Comment (laurentsimon) [reply internal/utils/marshall.go:58]: probably. I'll make the change. Thank you!

File: internal/utils/marshall.go
Comment (laurentsimon) [reply internal/utils/marshall.go:42]: yes!

**Review by asraa — APPROVED** [review]

**Comment by laurentsimon** [comment]
I'll wait for an ack from @ianlewis before merging.

File: internal/builders/go/pkg/config.go
Comment (ianlewis) [inline internal/builders/go/pkg/config.go:147]: nit: you can move these inside the if block below

File: internal/utils/path.go
Comment (ianlewis) [inline internal/utils/path.go:36]: nit: add a short godoc to these functions.

**Review by ianlewis — APPROVED** [review]

File: internal/builders/go/pkg/config.go
Comment (laurentsimon) [reply internal/builders/go/pkg/config.go:147]: good catch. Done.

File: internal/utils/path.go
Comment (laurentsimon) [reply internal/utils/path.go:40]: done. Let's have pre-submit to catch these in the future if we can

Comment (ianlewis) [reply internal/utils/path.go:40]: Yeah, this should be captured under #450

**Pull request merged**
